src/01_getting_started/03_state_of_async_rust.md

```diff
@@ -1,24 +1,105 @@
+  Rust application.
+- Execution of async code, IO and task spawning are provided by "async
+  runtimes", such as Tokio and async-std. Most async applications, and some
+  async crates, depend on a specific runtime. See
+  ["The Async Ecosystem"](../02_execution/01_chapter.md) section for more
+  details.
+
+Some language features you may be used to from regular Rust are not yet
+available in async Rust. Notably, Rust does not let you declare async
+functions in traits. Instead, you need to use workarounds to achieve the same
+result, which can be more verbose.
+
+## Compiling and debugging
+
+For the most part, compiler- and runtime errors in async Rust work
+the same way as they have always done in Rust. There are a few
+noteworthy differences:
+
+### Compilation errors
+
+Compilation errors in async Rust conform to the same high standards as
+regular Rust, but since async Rust often depends on more complex language
+features, such as lifetimes and pinning, you may encounter these types of
+errors more frequently.
+
+### Runtime errors
+
+Whenever the compiler encounters an async function, it generates a state
+machine under the hood. Stack traces in async Rust typically contain details
+from these state machines, as well as function calls from
+the runtime. As such, interpreting stack traces can be a bit more involved than
+it would be in regular Rust.
+
+### New failure modes
+
+A few novel failure modes are possible in async Rust, for instance
+if you call a blocking function from an async context or if you implement
+the `Future` trait incorrectly. Such errors can silently pass both the
+compiler and sometimes even unit tests. These pitfalls are easy to avoid through
+awareness.
+
+## Compatibility considerations
+
+Asynchronous and synchronous code cannot always be combined freely.
+For instance, you can't call an async function from a sync function.
+Sync and async code also tend to promote different design patterns, which can
+make it difficult to compose code intended for the different environments.
+
+Even async code cannot always be combined freely. Some crates depend on a
+specific async runtime to function. If so, it is usually specified in the
+crate's dependency list.
+
+These compatibility issues can limit your options, so make sure to thoroughly
+research which async runtime and what crates you may need as early as
+possible. Once you have settled in with a runtime, you won't have to worry
+about compatibility much.
+
+## Performance characteristics
+
+The CPU and memory overhead of async is typically significantly
+lower compared to using OS threads, especially when you have many lightweight
+tasks. Together with Rust's zero-cost abstractions, async Rust is one of the
+most performant concurrency environments available.
+The choice of an async runtime also impacts performance characteristics, but
+to a lesser extent.
+
+That said, async Rust can result in larger binary sizes, both due to the state
```

Comment:
Maybe some of what I said in the comment above should go in this paragraph

Reply:
Yes, I added specifically about IO bound workloads. In the future, we could have a section which is much more detailed about performance tradeoffs.